cmd/publicize/server.go

```diff
@@ -0,0 +1,237 @@
+	if err != nil {
+		return err
+	}
+	if !ok {
+		resp := fmt.Sprintf("only [%s](https://github.com/orgs/%s/people) org members may request a publicize", org, org)
+		logger.Info(resp)
+		return s.ghc.CreateComment(org, repo, num, plugins.FormatICResponse(ic.Comment, resp))
+	}
+
+	logger.Info("Publicize of PR has been requested.")
+
+	pr, err := s.ghc.GetPullRequest(org, repo, num)
+	if err != nil {
+		return err
+	}
+	baseBranch := pr.Base.Ref
+
+	var prCommits []string
+	commits, err := s.ghc.ListPRCommits(org, repo, num)
+	if err != nil {
+		return err
+	}
+
+	for _, commit := range commits {
+		prCommits = append(prCommits, commit.SHA)
+	}
+
+	// publicize only merged PRs.
+	if !pr.Merged {
+		resp := "cannot publicize an unmerged PR"
+		logger.Info(resp)
+		return s.ghc.CreateComment(org, repo, num, plugins.FormatICResponse(ic.Comment, resp))
+	}
+
+	logger.Infof("searching bond for %s/%s", org, repo)
+	bondedOrgRepo, ok := s.bonds[fmt.Sprintf("%s/%s", org, repo)]
+	if !ok {
+		logger.Warn("There is no bond for the current repository.")
+		resp := "cannot publicize because there is no bond for the current pull request"
+		return s.ghc.CreateComment(org, repo, num, plugins.FormatICResponse(ic.Comment, resp))
+	}
+
+	logger.Infof("Trying to merge the PR to destination: %s/%s@%s", org, repo, baseBranch)
+	if err := s.mergeAndPushToRemote(logger, bondedOrgRepo, num, org, repo, baseBranch, s.confirm, prCommits); err != nil {
+		return fmt.Errorf("couldn't merge the pull request and push to the destination: %v", err)
+	}
+
+	return s.ghc.CreateComment(org, repo, num, plugins.FormatICResponse(ic.Comment, "Publicize suceeded."))
+}
+
+func (s *server) mergeAndPushToRemote(logger *logrus.Entry, bondedOrgRepo string, prNumber int, org, repo, branch string, confirm bool, commits []string) error {
+	// this is already validated
+	bondedOrg := strings.Split(bondedOrgRepo, "/")[0]
+	bondedRepo := strings.Split(bondedOrgRepo, "/")[1]
+
+	bondURIRaw := fmt.Sprintf("https://github.com/%s/%s", bondedOrg, bondedRepo)
+	bondURI, err := url.Parse(bondURIRaw)
+	if err != nil {
+		logger.WithField("url", bondURIRaw).WithError(err).Error("Failed to construct URI")
+		return fmt.Errorf("failed to construct URI")
+	}
+	bondURI.User = url.User(string(s.githubTokenGenerator()))
+
+	prURIRaw := fmt.Sprintf("https://github.com/%s/%s", org, repo)
+	prURI, err := url.Parse(prURIRaw)
+	if err != nil {
+		logger.WithField("url", prURIRaw).WithError(err).Error("Failed to construct URI")
+		return fmt.Errorf("failed to construct URI")
+	}
+	prURI.User = url.User(string(s.githubTokenGenerator()))
+
+	gitDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return fmt.Errorf("Failed to create temporary directory for git operations: %v", err)
+	}
+
+	defer func() {
+		if err := os.RemoveAll(gitDir); err != nil {
+			logger.WithError(err).Error("Failed to clean up temporary directory")
+		}
+	}()
+
+	if err := gitExec(logger, gitDir, []string{"init"}...); err != nil {
```

Comment:
This belongs in a `git` client library?

Reply:
I can't use prow's git library for this.  Shall we extend it?

Reply:
Yes